### PR TITLE
Update SML example to use foldmap and other minor improvements

### PR DESCRIPTION
**Pull request by pclayton**
Here are some improvements to the SML example.

@fabjan - as you created the original example, please let us know whether you agree with the observations below.  One option is to add this as a separate example, but I think items 1 and 2 would still need to be addressed.  Note that the first commit addresses item in the original example.

 1. The code doesn't type check with SML/NJ.  Some explicit type annotations are required.  This is because SML/NJ uses a smaller scope than needed to resolve the types of flexible records.  (The Definition of SML is too lax about scope used to resolve flexible records, which is a known defect.)
 1. `xs @ [...]` is generally leads to O(n^2) behaviour because an entire new list is being created on each recursion.  This is avoided by extending the result of the inner recursion, which is achieved by storing the list in revserse.  However this is taken care of by `foldMapl` - see next comment.
 1. The nature of the problem is really a fold-map - a simultaneous fold and map.  Unfortunately the Basis Library did not provide such a combinator but most people write their own, so much so that it is now proposed for the Basis Library:
    https://github.com/SMLFamily/BasisLibrary/wiki/2015-003b-List
    The definition given above is as per the discussion section in that page.
 1. Once `foldMapl` is used, functions `positionLessons` and `positionSections` are not really needed anymore.
 1. Functions aren't mutually recursive, so separate fun bindings could be used.  I didn't change the code for this though.
 1. There is no code to evaluate `withPositions` so the code is not run.
 1. It's a good convention to avoid rebinding a function argument to avoid confusion.
    I've used `lessPos'1` instead of `lessPos`.  I tend to use a convention where
    successive versions of a value `x` have the name `x'<n>`, and the final version
    is `x'`.
 1. Inconsistent use of names: `start` vs. `lessPos`.
 1. Parentheses aren't required around an 'if' expression.
 1. For me it would be clearer to rename `setPosLessons` to `updateSection` and `setPos` to `updateLesson`.
 1. Inconsistent case convention of variables with `reset_lesson_position`.  (Note that record field names are variable names when they appear in a pattern, e.g. `val {a, ...} = ...`.)
 1. In most SML code I've seen/worked with, whitespace in record literals [usually](http://mlton.org/FunctionalRecordUpdate) have the form
    `{a = 1, b = 2}`
    rather than
    `{ a=1, b=2 }`.

### CODE REVIEW

**Comment by nickjj** [comment]
Hi, just popping in to say thanks for performing such an in depth analysis. I'll keep this open for a bit until the original creator gets a chance to look things over.

**Comment by fabjan** [comment]
Whew, that's a lot of input. Thanks for spreading more SML knowledge!

(1) My bad, I must have missed checking with SML/NJ for some of my last refactors.

(2) Yes, this was on purpose because the README says to prefer readability over code golfing (I extrapolated performance from that, perhaps that is not the intention). In my first versions I was consing to acc and then reving the list after the folds.

(3) I am not an SML expert and I did not know `foldMapL` was such a common idiom, but as I was writing it I was definitely looking for something like a fold-and-map in the ('97) basis library.

(4) They are not really needed with `foldl` either, but I think they might add to the readability. Do you think it's just too many too small functions to be idiomatic?

(5) I wrote the functions sort of top down instead of bottom down, so the `and`s are just there to avoid "unbound variable" errors. Perhaps this is not idiomatic either.

(6) Yeah, the README is not really clear about how things should execute. I just left that for other code outside.

(7) Oh, yeah I kind of agree. Not sure why I went this way here. I personally have been going `x`, `x'`, `x''`, ... and so on. I guess `x'4` is a bit nicer than `x''''`, and easier to tell apart if you for some reason use several versions in the same expression.

(10) OK, yeah they even had those names for a while.

(11) Yeah, this was to "comply" with the task at hand,. perhaps to strictly interpreted. 

(8, 9, 12) 👍 

In conclusion: I'm open to just replace the old version with this new solution, if `foldMapL` is as common as you say for people to write and use. We could keep two solutions (fixing up the first one with the feedback above) so there is one just using the basis library from '97, but then the combinator is so simple and useful so there is not that much use in keeping the two approaches, _I think_.

**Comment by nickjj** [comment]
I'll let you both determine the outcome.

I'm ok with either having 2 separate solutions or merging it into 1. Whatever you both think is best.

**Comment by fabjan** [comment]
Alright, do you have any thoughts on how strictly to interpret the description otherwise @nickjj? E.g. should we prefer to use idiomatic naming conventions in each language over what's in the README?

**Comment by nickjj** [comment]
I would use idiomatic naming conventions based on the implementation. File names like for-loop, map, foldl, etc. all come to mind as reasonable file name examples -- or use underscores instead of hyphens if that's what's idiomatic for your language.

If you need to break out into compound names that's ok. For example with Elixir there's a few like chunk_while_pattern_matching.

**Comment by fabjan** [comment]
@pclayton I incorporated your feedback on the original (and made https://github.com/fabjan/nested-data-structure-traversal/pull/1 to have a place to see the diff before opening up yet another PR here).

On the other hand I think the two solutions are very close though so it seems unnecessary to keep my original `foldl` solution, if using `foldMapL` is the way people would do it. I'll have to defer to you, I have not used SML outside of an educational context and that was 15 years ago.

File: sml/foldmapl_immutable_records.sml [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+and positionSection (section as {title, resetLessonPosition, lessons, ...}, (secPos, lessPos)) =
+  let
+    val lessPos'1 = if resetLessonPosition then 1 else lessPos
+    val (lessons', lessPos') = foldMapl positionLesson lessPos'1 lessons
+  in
+    (updateSection secPos lessons' section, (secPos + 1, lessPos'))
+  end
+
+and positionLesson (lesson, lessPos) = (updateLesson lessPos lesson, lessPos + 1)
+;
```
Comment (fabjan) [inline sml/foldmapl_immutable_records.sml:53]: Are these (the one after `sections'` too) extra semi colons needed?

File: sml/foldmapl_immutable_records.sml [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+ * `foldMapl` is proposed for the SML Basis Library - see
+ * https://github.com/SMLFamily/BasisLibrary/wiki/2015-003b-List
+ * The definition below is from the section Discussion.
+ *)
+fun foldMapl f init xs =
+  let
+    fun f' (x, (ys, r)) = let val (y, r) = f (x, r) in (y :: ys, r) end
+    val (ys, r) = List.foldl f' ([], init) xs
+  in
+    (List.rev ys, r)
```
Comment (fabjan) [inline sml/foldmapl_immutable_records.sml:38]: Is it praxis to refer to rev and foldl using the List structure like this in SML? Would you do this with everything in the top-level environment?

File: sml/foldmapl_immutable_records.sml (hunk 1 above)
Comment (pclayton) [reply sml/foldmapl_immutable_records.sml:53]: They are not needed.  A semicolon at the top-level ends a top-level declaration which means everything up to that semicolon is evaluated before considering the rest of the program, at least when submitting to a compiler running a REPL.  This ensures that all code up to the semicolon does not depend on anything following the semicolon, in particular there is no dependence on the subsequent code to resolve flexible records or overloading.  I added the semicolon out of principle, to make it clear that each section of code compiles without any subsequent code in scope.  This example is so simple that it doesn't add anything and it's probably simplest to drop the semicolons to avoid confusion.  I've dropped them in my updated branch.

File: sml/foldmapl_immutable_records.sml (hunk 2 above)
Comment (pclayton) [reply sml/foldmapl_immutable_records.sml:38]: For functions in the top-level environment that could apply to many types of data structure, e.g. `map`, `app`, `length`, `foldl` etc., it becomes habitual to qualify these with the structure for various reasons, for example

- the top-level variant may be hidden by another definition
- for clarity in code that uses the same function name with different types of data structure, e.g. `Vector.map` and `List.map`

Certainly, this is _not_ done for everything in the top-level environment.  I don't think I've ever seen anyone write `Bool.not`.  Also, the infix form of an operator can't be qualified by a structure.

**Comment by pclayton** [comment]
It is probably worth keeping the two solutions, on reflection.  I've updated my branch to update your `foldl` example based on the comments above, which has a series of commits resulting in something very similar to https://github.com/fabjan/nested-data-structure-traversal/pull/1 .  Then the last commit adds the `foldMapl` example.

File: sml/foldl_immutable_records.sml [hunk 3]
```diff
@@ -24,20 +24,20 @@ val sections : section list = [
     ]
 ]
 
-fun withPositions (sections : section list) =
-    (#1 o positionSections) sections
 
-and positionSections sections =
-    List.foldl positionSection ([], 1, 1) sections
+fun withPositions sections =
+    let val (revSections, _, _) = List.foldl positionSection ([], 1, 1) sections
```
Comment (fabjan) [inline sml/foldl_immutable_records.sml:29]: 👍 I like the extra clarity of prefixing the name with rev

**Review by fabjan — APPROVED** [review]
Good stuff!

Keeping the two solutions I think makes it easier to appreciate how something a simple difference as an added `foldMap` combinator can make the code more readable. And my understanding of this repository is that it's supposed to showcase different approaches.

**Comment by pclayton** [comment]
@fabjan Thanks.  Hope it's not another 15 years until you use SML again :)

@nickjj We're both happy, so this can be merged!

**Comment by nickjj** [comment]
Very nice, thanks for making this easy to merge and for the great discussion around SML.

**Comment by fabjan** [comment]
@pclayton one can dream!
